kubernetes/csi/lib/ubi_csi/controller_service.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+require "grpc"
+require "securerandom"
+require_relative "../csi_services_pb"
+require "logger"
+require "json"
+
+module Csi
+  module V1
+    class ControllerService < Controller::Service
+      MAX_VOLUME_SIZE = 2 * 1024 * 1024 * 1024 # 2GB in bytes
+      LOGGER = Logger.new($stdout)
+
+      def initialize
+        @volume_store = {} # Maps volume name to volume details
+        @mutex = Mutex.new
+      end
+
+      def log_with_id(id, message)
+        LOGGER.info("[req_id=#{id}] #{message}")
+      end
+
+      def run_cmd(*cmd, req_id: nil)
+        log_with_id(req_id, "Running command: #{cmd}") unless req_id.nil?
+        Open3.capture2e(*cmd)
+      end
+
+      def controller_get_capabilities(req, _call)
+        req_id = SecureRandom.uuid
+        log_with_id(req_id, "controller_get_capabilities request: #{req.inspect}")
+        if req.nil?
+          raise GRPC::InvalidArgument.new("Request cannot be nil", GRPC::Core::StatusCodes::INVALID_ARGUMENT)
+        end
+        resp = ControllerGetCapabilitiesResponse.new(
+          capabilities: [
+            ControllerServiceCapability.new(
+              rpc: ControllerServiceCapability::RPC.new(
+                type: ControllerServiceCapability::RPC::Type::CREATE_DELETE_VOLUME
+              )
+            )
+          ]
+        )
+        log_with_id(req_id, "controller_get_capabilities response: #{resp.inspect}")
+        resp
+      end
+
+      def select_worker_topology(req)
```

Comment:
some context in the form of comments around these functions could be helpful, like when, why and by whom these functions are called.